pkg/gcp/machine_controller_util.go

```diff
@@ -83,33 +79,31 @@ func (ms *MachinePlugin) CreateMachineUtil(ctx context.Context, machineName stri
 		instance.Description = *providerSpec.Description
 	}
 
-	disks := []*compute.AttachedDisk{}
+	var disks []*compute.AttachedDisk
 	for _, disk := range providerSpec.Disks {
-		var attachedDisk compute.AttachedDisk
-		autoDelete := false
-		if disk.AutoDelete == nil || *disk.AutoDelete == true {
-			autoDelete = true
+		attachedDisk := compute.AttachedDisk{
```

Comment:
why would you remove the else part? This now causes re-initialization of the `attachedDisk` if `disk.Type == api.GCPDiskTypeScratch` which is really not necessary.

Reply:
In my view, the enhanced readability achieved by reducing if-else conditions is more significant than the associated overhead of re-initialization.

Reply:
You can do the following:
```go
func createAttachedDisks(disks []*api.GCPDisk, zone, machineName string) []*compute.AttachedDisk {
	attachedDisks := make([]*compute.AttachedDisk, 0, len(disks))
	for _, disk := range disks {
		var attachedDisk compute.AttachedDisk
		switch disk.Type {
		case api.GCPDiskTypeScratch:
			attachedDisk = compute.AttachedDisk{
				Type:       api.GCPDiskTypeScratch,
				Boot:       false,
				AutoDelete: ptr.Deref(disk.AutoDelete, true),
				Interface:  disk.Interface,
				InitializeParams: &compute.AttachedDiskInitializeParams{
					DiskType: fmt.Sprintf("zones/%s/diskTypes/%s", zone, "local-ssd"),
				},
			}
		default:
			attachedDisk = compute.AttachedDisk{
				Type:       api.GCPDiskTypePersistent,
				Boot:       disk.Boot,
				AutoDelete: ptr.Deref(disk.AutoDelete, true),
				InitializeParams: &compute.AttachedDiskInitializeParams{
					DiskSizeGb:            disk.SizeGb,
					DiskType:              fmt.Sprintf("zones/%s/diskTypes/%s", zone, disk.Type),
					Labels:                disk.Labels,
					SourceImage:           disk.Image,
					ProvisionedIops:       disk.ProvisionedIops,
					ProvisionedThroughput: disk.ProvisionedThroughput,
				},
			}
		}

		if disk.Encryption != nil {
			attachedDisk.DiskEncryptionKey = &compute.CustomerEncryptionKey{
				KmsKeyName:           strings.TrimSpace(disk.Encryption.KmsKeyName),
				KmsKeyServiceAccount: strings.TrimSpace(disk.Encryption.KmsKeyServiceAccount),
			}
			klog.V(3).Infof("(CreateMachineUtil) For machineName: %q, diskLabel: %q, DiskEncryptionKey.KmsKeyName: %q, "+
				"DiskEncryptionKey.KmsKeyServiceAccount: %q",
				machineName,
				disk.Labels["name"],
				attachedDisk.DiskEncryptionKey.KmsKeyName,
				attachedDisk.DiskEncryptionKey.KmsKeyServiceAccount)
		}
		attachedDisks = append(attachedDisks, &attachedDisk)
	}
	return attachedDisks
}
```

Then just call this function from `CreateMachineUtil`